llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/Vectorization.cpp

```diff
@@ -0,0 +1,258 @@
+    return input;
+  }
+
+  Location loc = packOp.getLoc();
+  ShapedType inputType = packOp.getInputType();
+  int64_t inputRank = inputType.getRank();
+
+  SmallVector<int64_t> paddedShape;
+  DenseMap<int64_t, OpFoldResult> tileAndPosMapping =
+      packOp.getDimAndTileMapping();
+  for (int64_t dim = 0; dim < inputRank; ++dim) {
+    int64_t size = inputType.getDimSize(dim);
+    if (!tileAndPosMapping.count(dim)) {
+      paddedShape.push_back(size);
+      continue;
+    }
+
+    Optional<int64_t> tileSize =
+        getConstantIntValue(tileAndPosMapping.lookup(dim));
+    assert((!inputType.isDynamicDim(dim) && tileSize.hasValue()) &&
+           "something goes really wrong...");
+    int64_t sizeWithPad = llvm::alignTo(size, tileSize.getValue());
+    paddedShape.push_back(sizeWithPad);
+  }
+  auto resultType =
+      RankedTensorType::get(paddedShape, inputType.getElementType());
+  return tensor::createPadHighOp(resultType, input, packOp.getPaddingValue(),
+                                 /*nofold=*/false, loc, builder);
+}
+
+// Creates a linalg.generic that transposes input using permutation indices.
+// Example: (M1, M0, N1, N0) -> (M1, N1, M0, N0) if indices = {0, 2, 1, 3}.
+Value createTransposeOp(Location loc, OpBuilder &builder, Value input,
+                        ArrayRef<int64_t> indices) {
+  auto inputType = input.getType().cast<RankedTensorType>();
+  auto nloops = indices.size();
+
+  SmallVector<AffineExpr> exprs = llvm::to_vector<4>(
+      llvm::map_range(indices, [&](int64_t index) -> AffineExpr {
+        return builder.getAffineDimExpr(index);
+      }));
+
+  ArrayRef<int64_t> inputShape = inputType.getShape();
+  SmallVector<OpFoldResult> targetShape;
+  for (auto i : indices) {
+    if (inputShape[i] == ShapedType::kDynamicSize) {
+      targetShape.emplace_back(builder.create<tensor::DimOp>(loc, input, i));
+    } else {
+      targetShape.push_back(builder.getIndexAttr(inputShape[i]));
+    }
+  }
+
+  Value empty = builder.create<tensor::EmptyOp>(loc, targetShape,
+                                                inputType.getElementType());
+  SmallVector<StringRef, 4> loopAttributeTypes(nloops,
+                                               getParallelIteratorTypeName());
+  SmallVector<AffineMap, 2> indexingMaps = {
+      inversePermutation(
+          AffineMap::get(nloops, 0, exprs, builder.getContext())),
+      AffineMap::getMultiDimIdentityMap(nloops, builder.getContext())};
+
+  auto transposedOp = builder.create<linalg::GenericOp>(
+      loc, empty.getType(), input, empty, indexingMaps, loopAttributeTypes,
+      [&](OpBuilder &nestedBuilder, Location nestedLoc, ValueRange args) {
+        nestedBuilder.create<linalg::YieldOp>(nestedLoc, args[0]);
+      });
+
+  return transposedOp.getResult(0);
+}
+
+// Rewrites iree_linalg_ext.pack to tensor.pad + tensor.expand_shape +
+// linalg.generic (transpose) ops.
+struct GeneralizePackOpPattern : OpRewritePattern<PackOp> {
```

Comment:
IMO we should avoid return failure if any of the outer dims is != 1.
I see you changed tests recently to all 1's.

Once we only accept leading 1's dimension, we should avoid reshapes and instead use rank-reducing insert/extract.

Then there would be no need for new vectorization patterns.